README.md
# EventNotifier

[![Pub Package](https://img.shields.io/pub/v/eventnotifier.svg?style=flat-square)](https://pub.dev/packages/eventnotifier)

> NOTE: Consider using the [Event] package instead.

> This package is an implementation of a central 'Event Bus' or 'Event Broker' which is suited to some specific use-cases.

>In general, using a lightweight custom [Event] is a preferable way to have something that is independent of dependencies, inform others that something of interest has occurred.

Broadcasts named events to interested subscribers. When an event occurs, a method (callback) associated with the subscriber is executed.

EventNotifier is an implementation of the Observer Pattern, providing the ability to publish events, and subscribe to them elsewhere. Such implementations are sometimes called an 'Event Broker' or 'Event Bus'.

This is primarily intended to be mixed-in with a problem domain model, to enable it to provide change notifications.

## Background

It can be ideal to model problems or systems independent of _user interfaces_ (UI) or _system interfaces_ (SI). Doing so, one can model the problem or system without regard to how the model may (or may not be) consumed.

However, given an independent model, how can something else outside it know if something in the domain model changed?  One solution is to use `EventNotifier`.

### An Illustration

Consider that one wanted to model the operation of a single 'Elevator'. One might produce an independent domain model in object-oriented Dart code that modeled it.

Later, one might want to provide a user interface (UI) that let one interact with the model, or provide a system interface (SI) that interacted with (for example) the [Programmable Logic Controller (PLC)][plc] of a real-life Elevator.

In order to let external modules or systems know that the Elevator changed _floors_, the Elevator model can notify them through `EventNotifier`, so they can react as appropriate.

## See also

[Event] - Create lightweight custom Dart Events. An alternative to this centralized Event Bus/Event Broker approach.

## Dependencies

None. This Dart package has no non-development dependencies on any other packages.

## Usage

A simple example (see the  static EventNotifier test method):

```dart
import 'package:eventnotifier/eventnotifier.dart';

main() {
    var e = EventNotifier();

    e.subscribe('myChange', (_) => print('boom'));
    e.subscribe('myChange', (args) => print(args['pi']));

    // number of event names
    assert(e.count() == 1);
    // number of handlers subscribed to the event name
    assert(e.count('myChange') == 2);

    // indicate that event 'myChange' has occured
    // rguments are only required if a subscriber
    // expects one or more
    e.notify('myChange', {'pi': 3.14159});

    // displays ...
    // boom
    // 3.14159
}
```

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/aryehof/eventnotifier/issues
[event]: https://pub.dev/packages/event
[plc]: https://en.wikipedia.org/wiki/programmable_logic_controller


CHANGELOG.md
# Changelog

## Version 1.1.1  (2020-01-23)

- Added recommendation to generally use the [Event] package to the README.
- Added licence header to source files.

## Version 1.1.0  (2020-01-23)

- Removed references to EventSubscriber, as it now supports [Event] rather than EventNotifier


## Version 1.0.5  (2020-01-14)

- Add support again for optional event arguments (was previously removed).
Arguments should now be supplied as a Map, e.g. notify('myEvent', {'age': 32});
- A more detailed error is shown if a subscriber expects an argument, but
it isn't supplied when notified.
- One can now access the EventNotifier using the getEventNotifier() method.
Getting the instance can be useful from within a mixin.

## Version 1.0.4  (2020-01-13)

- Fixed LICENSE reference to package

## Version 1.0.3  (2020-01-13)

- Package example updated to show the static test() method.
- Add a link to EventSubscriber in the README.

## Version 1.0.2  (2020-01-12)

- Removed support for optional event arguments. Query the model instead.
- Exceptions are now also output to the console.
- README describes how to use Flutter with EventNotifier

## Version 1.0.1  (2020-01-11)

- Improved README and minor reformatting changes

## Version 1.0.0  (2020-01-11)

- Initial release

[event]: https://pub.dev/packages/event


test/event_broker_test.dart
import 'package:event_broker/event_broker.dart';
import 'package:test/test.dart';

void main() {
  group('A group of tests', () {
    setUp(() {
      // awesome = Awesome();
    });

    test('First Test', () {
      expect(() => EventBroker.test(), returnsNormally);
    });
  });
}


test/eventnotifier_test.dart
import 'package:eventnotifier/eventnotifier.dart';
import 'package:test/test.dart';

void main() {
  group('EventNotifier', () {
    test('Static test', () {
      expect(() => EventNotifier.test(), returnsNormally);
    });
  });
}


lib/src/error.dart
// Copyright 2020 Aryeh Hoffman. All rights reserved.
// Use of this source code is governed by an Apache-2.0 license that can be
// found in the LICENSE file.

class NotifyException implements Exception {
  // The error message to display
  String _message;

  /// The program location of the error, e.g. the method name (optional)
  String _errorLocation;

  NotifyException(String message, [String errorLocation]) {
    _message = message;
    _errorLocation = errorLocation;
  }

  @override
  String toString() {
    var fullMessage;

    if (_errorLocation == null) {
      // no location specified
      fullMessage = 'Error (EventNotifier): ' + _message;
    } else {
      fullMessage = 'Error (EventNotifier:$_errorLocation): ' + _message;
    }

    print('message');
    return fullMessage;
  }
}


lib/src/event_broker_base.dart
import 'dart:async';

typedef VoidCallback = void Function(List<dynamic> args);

class EventBroker {
  final _myMap = <String, List<VoidCallback>>{};
  int _version = 0;
  int _microtaskVersion = 0;

  /// register an event
  void register(String name) {
    if (name.isEmpty) throw Exception('Error (register): a name is required to register an event');
    if (_myMap.containsKey(name))
      throw Exception('Error (register): attempted to register a name that already exists');

    _myMap[name] = [];
  }

  /// subscribe to a registered event, with the provided callback
  void subscribe(String name, VoidCallback callback) {
    if (!_myMap.containsKey(name))
      throw Exception('Error (subscribe): provided name is not registered');
    if (callback == null) throw Exception('Error (subscribe): a callback is required');

    _myMap[name].add(callback);
  }

  /// publish an event with optional arguments
  void publish(String name, [List<dynamic> args]) {
    if (!_myMap.containsKey(name))
      throw Exception('Error (publish): publish name ("$name") not found');

    // We schedule a microtask to debounce multiple changes that can occur
    // all at once.
    if (_microtaskVersion == _version) {
      _microtaskVersion++;
      scheduleMicrotask(() {
        _version++;
        _microtaskVersion = _version;

        // Convert the Set to a List before executing each listener. This
        // prevents errors that can arise if a listener removes itself during
        // invocation!
        // _listeners.toList().forEach((VoidCallback listener) => listener());
        _myMap[name].toList().forEach((VoidCallback callback) => callback(args));
      });
    }
  }

  void remove(String name) {
    _myMap.remove(name);
  }

  int get count => _myMap.length;

  static void test() {
    var e = EventBroker();
    e.register('mine');
    e.subscribe('mine', (_) => print('boom'));
    e.subscribe('mine', (args) => print(args[0]));
    assert(e.count == 1);
    e.publish('mine', [99]);
  }
}


lib/src/eventnotifier.dart
// Copyright 2020 Aryeh Hoffman. All rights reserved.
// Use of this source code is governed by an Apache-2.0 license that can be
// found in the LICENSE file.

import 'dart:async';
import 'package:eventnotifier/src/error.dart';

typedef NotificationCallback = void Function(Map<String, dynamic> args);

/// Broadcasts named events to interested subscribers.
/// When an event occurs, a method (callback) associated with the subscriber is executed.
class EventNotifier {
  final _eventMap = <String, List<NotificationCallback>>{};
  int _version = 0;
  int _microtaskVersion = 0;

  /// Subscribe the provided method (callback) to the specified named event.
  void subscribe(String eventName, NotificationCallback method) {
    if (eventName.isEmpty) throw NotifyException('no eventName provided', 'subscribe');
    if (method == null) throw NotifyException('a method (callback) is required', 'subscribe');

    if (_eventMap.containsKey(eventName)) {
      _eventMap[eventName].add(method);
    } else {
      _eventMap[eventName] = [method];
    }
  }

  /// Notify each subscriber that the specified named event has occured.
  /// An optional map of arguments can be attached to the notification
  /// e.g. myEventNotifier.notify('valueChanged', {'age': 32);
  void notify(String eventName, [Map<String, dynamic> args]) {
    // ignore if no subscribers
    if (!_eventMap.containsKey(eventName)) return;

    // Schedule a microtask to debounce multiple changes that can occur all at once.
    if (_microtaskVersion == _version) {
      _microtaskVersion++;
      scheduleMicrotask(() {
        _version++;
        _microtaskVersion = _version;

        // Convert the Set to a List before executing each callback. This
        // prevents errors that can arise if a callback removes itself during invocation
        try {
          _eventMap[eventName].toList().forEach((NotificationCallback callback) => callback(args));
        } on NoSuchMethodError {
          throw NotifyException(
              'an argument expected by a subscriber, was not included in the notification for event "$eventName"',
              'notify');
        }
      });
    }
  }

  /// Remove the named event's callback
  void remove(String eventName, NotificationCallback callback) {
    if (!_eventMap.containsKey(eventName))
      throw NotifyException('event name ("$eventName") not found', 'remove');

    _eventMap[eventName].remove(callback);
  }

  /// Gets the number of named events if no eventName is specified.
  /// If an eventName is specified, returns the number of subscribed handlers for that named event.
  int count([String eventName]) {
    if (eventName == null) {
      return _eventMap.length;
    } else {
      if (!_eventMap.containsKey(eventName))
        throw NotifyException('specified eventName not found', 'count');
      return _eventMap[eventName].length;
    }
  }

  // Get  the EventNotifier instance. Useful when EventNotifier is mixed-in with a domain model.
  EventNotifier getEventNotifier() {
    return this;
  }

  /// Test functionality, run EventNotifier.test()
  static void test() {
    var e = EventNotifier();

    e.subscribe('myChange', (_) => print('boom'));
    e.subscribe('myChange', (args) => print(args['pi']));

    // number of event names
    assert(e.count() == 1);
    // number of handlers subscribed to the event name
    assert(e.count('myChange') == 2);

    // indicate that event 'myChange' has occured
    // rguments are only required if a subscriber
    // expects one or more
    e.notify('myChange', {'pi': 3.14159});

    // displays ...
    // boom
    // 3.14159
  }
}
